### ISSUE TITLE
Build for Ubuntu

### ISSUE THREAD

**Issue by AdelKS** [Frogging-Family/linux-tkg#21]
Hello!

First of all, thank you for this nice repository with select patches for Linux gaming. I hope I am not annoying with this issue. 

I want to use your repository to build a custom kernel for my Ubuntu 20.04 with Fsync and march=native optimisations with GCC 10.1. Unfortunately I couldn't find a way to make `makepkg -si` work (it seems that it's part of Arch's `pacman`). For Ubuntu, I found [this wiki page](https://wiki.ubuntu.com/KernelTeam/GitKernelBuild) to build from the Linux git tree and it works awesomely, too simple to be honest. I expected more work.

A possibility to do what I want is that I only apply the patches in the folder `linux57-tkg-patches` after a `git checkout v5.7` in the linux tree, but I feel like I will be missing some tweaks from the other files on your repo, that I suppose will tweak the `.config` file. 

Do you have any better ideas on how I can use your repository to build an Ubuntu kernel (that makes .deb packages) ?

Thank you!

Adel

**Comment by AdelKS** [comment]
I successfully compiled a fully working v5.7 kernel with your patches, I mainly followed the `PKGBUILD` script off by hand and chose what I wanted. I think I can derive a shell script that works on Ubuntu from that file, a very basic version first, would you be interested in integrating it in the repository ?

**Comment by Tk-Glitch** [comment]
It was talked about in https://github.com/Tk-Glitch/PKGBUILDS/issues/539 (which is basically a duplicate) and is still on my to-do list (with very low priority, I have to admit). The whole nightmare with wine right now has been slowing my other works down.

My plan is to have a core script for source preparation that would then get used by the PKGBUILD and, as/if needed, other scripts (in a similar fashion to wine/proton-tkg) for portability sakes.

A Ubuntu-centric script would definitely be of help.

**Comment by AdelKS** [comment]
Oh I haven't followed closely what's happening with wine, though I did clone it and to compiled the latest master version without thinking further than modding the `customisation.cfg` file. Good luck with all that !

> My plan is to have a core script for source preparation that would then get used by the PKGBUILD and, as/if needed, other scripts (in a similar fashion to wine/proton-tkg) for portability sakes.

I was actually thinking about that when I will do the Ubuntu script, since I think I will use a big part of the existing `PKGBUILD` as is, I will come up with a proposal for ya then (after checking out what you are doing with wine-tkg). Though I have very little knowledge of shell scripting, I was very tempted by doing a python one. But I will try nonetheless with shell and learn on the fly, let's see how that goes.

**Comment by AdelKS** [comment]
@Tk-Glitch I have a few questions for you, as I am working on the script :
- Do you think it's generally safe and better to start off the distro's own `.config` file then apply the patches on it, when it's not Arch ? That's what I did when I manually compiled my Ubuntu custom Kernel. I discovered that the script uses the config file `config.x86_64` in the folder `linux-5x-tkg-config`
- Isn't it better to just take the compressed folder `https://www.kernel.org/pub/linux/kernel/v5.x/linux-${pkgver}.tar.xz` instead of downloading the base version then patching it ? (line 100 in PKGBUILD)
- Is it okay to move the initial commands in PKGBUILD (line 26 to 153) and move them into a function, for example `_init_script()` in the `prepare` file ? Then use that function in both PKGBUILD and the script I am writing ?

**Comment by Tk-Glitch** [comment]
1. Yes. Specific settings may be expected on the distro's side so preserving default config should give optimal compat.

2. Better is subjective. It's a very small design choice that leads to less bandwidth used for someone keeping the tarballs around. Not something really important for most people.

3. It's okay for use in an external script, but it's not ok for the PKGBUILD itself. There are quite a few rules regarding PKGBUILDs formatting and that part is more or less untouchable and has to stay there for makepkg to accept running it. For wine-tkg's distro-agnostic script I have decided to re-implement that part while mimicking the expected behavior of makepkg to preserve compatibility with repository clones it creates etc. That is not a requirement here since it's limited to tarballs handling, which simplifies things on that front.

4. It's absolutely okay for anything that won't run from the PKGBUILD. `plain ` is used by makepkg and has no real use in linux-tkg scripts outside of wrapping echo. A `here-document ` redirection is usually what I'm using for the frog header, for example:
```
cat << 'EOM'
       .---.`               `.---.
    `/syhhhyso-           -osyhhhys/`
   .syNMdhNNhss/``.---.``/sshNNhdMNys.
   +sdMh.`+MNsssssssssssssssNM+`.hMds+
   :syNNdhNNhssssssssssssssshNNhdNNys:
    /ssyhhhysssssssssssssssssyhhhyss/
    .ossssssssssssssssssssssssssssso.     (Mostly)
   :sssssssssssssssssssssssssssssssss:    Portable
  /sssssssssssssssssssssssssssssssssss/        GCC/MingW
 :sssssssssssssoosssssssoosssssssssssss:
 osssssssssssssoosssssssoossssssssssssso
 osssssssssssyyyyhhhhhhhyyyyssssssssssso
 /yyyyyyhhdmmmmNNNNNNNNNNNmmmmdhhyyyyyy/
  smmmNNNNNNNNNNNNNNNNNNNNNNNNNNNNNmmms
   /dNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNd/
    `:sdNNNNNNNNNNNNNNNNNNNNNNNNNds:`
       `-+shdNNNNNNNNNNNNNNNdhs+-`
             `.-:///////:-.`

EOM
```

"Plain" `echo` can break PKGBUILDs in various ways but a simple wrap should be enough to prevent problems.
Here's what I did for the re-implementation of the active echoing funcs of makepkg, which is pretty basic:
```
# use msg2, error and pkgver funcs for compat
msg2() {
 echo -e " \033[1;34m->\033[1;0m \033[1;1m$1\033[1;0m" >&2
}
error() {
 echo -e " \033[1;31m==> ERROR: $1\033[1;0m" >&2
}
warning() {
 echo -e " \033[1;33m==> WARNING: $1\033[1;0m" >&2
}
```

Hope it helps

**Comment by AdelKS** [comment]
1. Got it!
2. Got it!
3. Got it! So I will copy it and strip it down to what I need. I will also work on a Fedora version. Since everything is the same except the make and install commands. 
4. Got it!

I am also about thinking to enable uninstalling through the script for people who wouldn't know how to : I haven't looked up how to uninstall a kernel if it is installed with a `make install`, which is Fedora's case. I guess it is `make uninstall`, but that would require to keep all the files untouched. Any thoughts on this ?

**Comment by AdelKS** [comment]
The install script has been merged in #26 ! This issue can be closed now :smile:
